boards: arm: nrf5340_dk: Enable DC/DC by default & Fix NRF_CACHE and NVMC macro defines

cvinayak opened the pull request:
Since the nRF5340 DK contains the required DC/DC circuitry,
enable it by default while keeping it configurable. This
reduces power consumption in application core , network
core and high voltage use.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Fixes #20695

cvinayak committed:
boards: arm: nrf5340_dk: Enable DC/DC by default

Since the nRF5340 DK contains the required DC/DC circuitry,
enable it by default while keeping it configurable. This
reduces power consumption in application core , network
core and high voltage use.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ioannisg reviewed (changes requested):
Thanks for the patch @cvinayak .
Note however it won't work for App mcu executing on non secure mode
We need to have a way to access regulators from NS domain or just don't do it at all.

cvinayak commented:
I just followed the existing code for NRF_CACHE_S, how different should it be that soc.c can handle NS mode? Feel free to suggest changes.

ioannisg commented:
> I just followed the existing code for NRF_CACHE_S, how different should it be that soc.c can handle NS mode? Feel free to suggest changes.

Oh, right, I understand your confusion. It should not be different from CACHE, this seems to be a minor issue IMO.

So, `SOC_NRF5340_CPUAPP `is selected for both Secure and Non-Secure builds for the Application MCU. Therefore, any access to _S needs to be guarded not only by `CONFIG_SOC_NRF5340_CPUAPP`, but also with `!CONFIG_TRUSTED_EXECUTION_NONSECURE`.

The reason it works for nRF53 is that the whole thing is guarded by CONFIG_NRF_ENABLE_CACHE, and _that_ is defined as.
```
config NRF_ENABLE_CACHE
	bool "Enable cache"
	depends on (SOC_NRF5340_CPUAPP && !TRUSTED_EXECUTION_NONSECURE) \
			|| SOC_NRF5340_CPUNET
	default y
```
...i.e. the guarding with !TRUSTED_EXECUTION_NONSECURE occurs already at the Kconfig definition.

The reason I did it this way is that I did not want the `NRF_ENABLE_CACHE `option to be enabled at all for Non-Secure builds.

To summarize. what I suggest is to guard your REGULATORS_S aditionally with !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE) and if you want, you can add the same guard around CACHE_S.

ioannisg reviewed (changes requested):
Actually, please, disregard my comment, @cvinayak seems we have a way to pull the right NRF_xxx macro via nrfx, now (enlighted by @anangl ).
I think you can simply use NRF_REGULATORS, then. Not sure if this is present on Network MCU.

ioannisg commented on boards/arm/nrf5340_dk_nrf5340/Kconfig:
```diff
@@ -5,6 +5,21 @@
 if BOARD_NRF5340_DK_NRF5340_CPUAPP || BOARD_NRF5340_DK_NRF5340_CPUAPPNS
 
+config BOARD_ENABLE_DCDC_APP
+	bool "Enable Application Core DCDC converter"
```
Please, use "MCU" instead of "Core" ; this is the name we use consistently in doc/index.rst.

ioannisg replied:
and in similar places below

ioannisg commented on soc/arm/nordic_nrf/nrf53/Kconfig.soc:
```diff
@@ -89,6 +89,21 @@ config SOC_NRF5340_CPUNET_QKAA
+config SOC_DCDC_NRF53X_APP
+	bool
+	help
+	  Enable nRF53 series System on Chip Application Core DC/DC converter.
```
PLease, use "MCU" instead of Core, here, and below

ioannisg commented on soc/arm/nordic_nrf/nrf53/Kconfig.soc:
```diff
@@ -89,6 +89,21 @@ config SOC_NRF5340_CPUNET_QKAA
 
 endchoice
 
+config SOC_DCDC_NRF53X_APP
```
Can you add dependencies in these options, i.e. depend on SOC_NRF5340_CPUAPP

cvinayak replied:
done

ioannisg commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -59,6 +59,18 @@ static int nordicsemi_nrf53_init(struct device *arg)
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP)
+#if defined(CONFIG_SOC_DCDC_NRF53X_APP)
+	NRF_REGULATORS_S->VREGMAIN.DCDCEN = 1;
```
Just try to use NRF_REGULATORS here, i.e. without _S

ioannisg replied:
and below

ioannisg committed:
boards: arm: nrf5340_dk: fix nRF Cache and NVMC macro usage

We need to fix the formato of the macros for nRF CACHE and NVMC
peripherals in soc.c, so the _S and _NS suffixes are skipped.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg reviewed (commented):
@anangl could you take a look as well?

anangl commented on soc/arm/nordic_nrf/nrf53/soc.c:
```diff
@@ -59,6 +59,18 @@ static int nordicsemi_nrf53_init(struct device *arg)
 	*((u32_t *)0x500046D0) = 0x1;
 #endif
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP)
```
This does not seem to be needed.
